src/main/dx.ts

```diff
@@ -0,0 +1,423 @@
+ * @param allProjects {boolean} Should we limit to St. Jude Cloud projects
+ *   or list all projects?
+ * @param cb {SuccessCallback<IProject[]>} If successful, all projects meeting
+ *   the criteria of the options passed in.
+ */
+export function listProjects(
+  token: string,
+  allProjects: boolean,
+  cb: SuccessCallback<IProject[]>,
+) {
+  let tagsToCheck: string[] = [];
+
+  if (!allProjects) {
+    tagsToCheck = [config.TOOL_PROJECT_TAG, config.DATA_PROJECT_TAG];
+  }
+
+  const client = new Client(token);
+
+  const options = {
+    describe: {
+      fields: {
+        name: true,
+        level: true,
+      },
+    },
+    level: ProjectLevel.Upload,
+    tags: tagsToCheck.length > 0 ? { $or: tagsToCheck } : undefined,
+  };
+
+  client.system
+    .findProjects(options)
+    .then(({ results }: { results: IProject[] }) => {
+      cb(null, results);
+    })
+    .catch((err: any) => {
+      cb(err, null);
+    });
+}
+
+/**
+ * @TODO(mmacias)
+ *
+ * @param client {Client} A preexisting Client for interacting with the DNAnexus API.
+ * @param projectId {string} The DNAnexus project identifier (project-XXXXXXX).
+ */
+const fileUploadParameters = async (
+  client: Client,
+  projectId: string,
+): Promise<IFileUploadParameters> => {
+  const result = await client.project.describe(projectId, {
+    fields: { fileUploadParameters: true },
+  });
+
+  const { fileUploadParameters: params } = result;
+
+  if (!params) {
+    throw new Error('missing file upload parameters');
+  }
+
+  return params;
+};
+
+class UploadTransfer {
+  private client: Client;
+  private src: string;
+  private size: number;
+  private progressCb: ResultCallback<number>;
+  private finishedCb: SuccessCallback;
+
+  private request: Request | null = null;
+  private bytesRead = 0;
+
+  public constructor(
+    token: string,
+    src: string,
+    progressCb: ResultCallback<number>,
+    finishedCb: SuccessCallback,
+  ) {
+    this.client = new Client(token);
+    this.src = src;
+    this.size = fs.statSync(this.src).size;
+    this.progressCb = progressCb;
+    this.finishedCb = finishedCb;
+  }
+
+  public async prepare(projectId: string): Promise<utils.IByteRange[]> {
+    const { maximumPartSize } = await fileUploadParameters(
+      this.client,
+      projectId,
+    );
+
+    return utils.byteRanges(this.size, maximumPartSize);
+  }
+
+  public async start(projectId: string, dst: string) {
+    const ranges = await this.prepare(projectId);
+    const name = path.basename(this.src);
+
+    const { id } = await this.client.file.new({
+      folder: dst,
+      name,
+      parents: true,
+      project: projectId,
+      // tags: [config.NEEDS_ANALYSIS_TAG],
+    });
+
+    for (let i = 0; i < ranges.length; i++) {
+      const { start, end } = ranges[i];
+      await this.transfer(id, i + 1, start, end);
+    }
+
+    await this.client.file.close(id);
+
+    this.finishedCb(null, {});
```

Comment:
@zaeleus not sure if the correct type was returned here.

Reply:
It's `undefined`; there is no result. Nothing up the call stack uses it. The empty object is remnants of old code and left there since it's not optional.

Reply:
Okay, I can just remove it then.